core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -870,22 +874,21 @@ private[spark] class TaskSchedulerImpl(
   }
 
   /**
-   * Marks the task has completed in all TaskSetManagers for the given stage.
+   * Marks the task has completed in the active TaskSetManager for the given stage.
    *
    * After stage failure and retry, there may be multiple TaskSetManagers for the stage.
-   * If an earlier attempt of a stage completes a task, we should ensure that the later attempts
-   * do not also submit those same tasks.  That also means that a task completion from an earlier
-   * attempt can lead to the entire stage getting marked as successful.
+   * If an earlier zombie attempt of a stage completes a task, we can ask the later active attempt
+   * to skip submitting and running the task for the same partition, to save resource. That also
+   * means that a task completion from an earlier zombie attempt can lead to the entire stage
+   * getting marked as successful.
    */
-  private[scheduler] def markPartitionCompletedInAllTaskSets(
+  private[scheduler] def markPartitionCompleted(
```

Comment:
Doesn't this method need to acquire synchronization on `TaskSchedulerImpl` object?

Reply:
no it doesn't. It's called in the thread `task-result-getter`

Reply:
Correct me if I am wrong, but if I follow the call chain correctly, you are calling the method `markPartitionCompleted` from `enqueuePartitionCompletionNotification`, which is called from `notifyPartitionCompletion`, and which gets invoked inside the DAG Scheduler `event loop thread`. Am I missing something here?

Reply:
Ahh I see it now, thanks.

Reply:
taskResultGetter is a multithreaded pool (default 4) so I think you still need extra protection here

Reply:
ah good catch! I should do synchronization here.

Reply:
Just fyi, after adding the synchronized statement, this PR is almost  similar to PR #22806 before it got restructured. I could have simply added synchronized to `markPartitionCompletedInAllTaskSets()` instead of restructuring the PR but did not do so because of the following suggestion by @squito which I agree with.

> I'm not sure how to fix this. You could make TaskSchedulerImpl.markPartitionCompletedInAllTaskSets() synchronized, but then you're getting a lock on the taskSchedulerImpl in the DAGScheduler event loop. That's not good for scheduling throughput, and also want to make sure there is no change of deadlock.

So I would like to once again pose the same question as above: is the scheduling throughput getting impacted by adding the extra synchronization?

Reply:
Ah thanks for pointing it out! Yes it's very similar, except that this PR sends message to the task result getter thread pool first, to avoid locking in the DAGScheduler event loop thread.

I don't think we will hit a deadlock. The task result getter thread pool calls `TaskSchedulerImpl.handleSuccessfulTask`, which is synchronized too. And `handleSuccessfulTask` calls method of `TaskSetManager`, which is the same as the newly added `handlePartitionCompleted`.